### PR TITLE
lndclient: add (incomplete) QueryRoutes

### PR DESCRIPTION
**Pull request by bhandras**
#### Pull Request Checklist

- [x] PR is opened against correct version branch.
~- [ ] Version compatibility matrix in the README and minimal required version
      in `lnd_services.go` are updated.~
~- [ ] Update `macaroon_recipes.go` if your PR adds a new method that is called
  differently than the RPC method it invokes.~

### CODE REVIEW

File: lightning_client.go [hunk 1]
```diff
@@ -806,6 +811,80 @@ type NetworkInfo struct {
...
 
+// QueryRoutesRequest is the request of a QueryRoutes call.
+type QueryRoutesRequest struct {
+	// Source is the optional source vertex of the route.
+	Source *route.Vertex
+
+	// PubKey is the destination vertex.
+	PubKey route.Vertex
+
+	// LastHop is the optional last hop befor the destination.
```
Comment (carlaKC) [inline lightning_client.go:822]: s/befor/before

File: lightning_client.go [hunk 2]
```diff
@@ -806,6 +811,80 @@ type NetworkInfo struct {
...
+	LastHop *route.Vertex
+
+	// RouteHints represents the different routing hints that can be used to
+	// assist the router. These hints will act as optional intermediate hops
+	// along the route.
+	RouteHints [][]zpay32.HopHint
```
Comment (carlaKC) [inline lightning_client.go:828]: What does "optional intermediate hops" mean? I think we only supply hints for the final hop? 
Nitty though, I guess in theory we could have hints for other private channels that don't belong to the last hop?

File: lightning_client.go [hunk 3]
```diff
@@ -3098,3 +3177,88 @@ func (s *lightningClient) ListPermissions(
...
+		return nil, err
+	}
+
+	resp, err := s.client.QueryRoutes(rpcCtx, rpcReq)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Routes) == 0 {
+		return nil, fmt.Errorf("no route found")
```
Comment (carlaKC) [inline lightning_client.go:3246]: Can we define this error so that callers don't need to string match?

File: lightning_client.go [hunk 4]
```diff
@@ -3098,3 +3177,88 @@ func (s *lightningClient) ListPermissions(
...
+	}
+
+	route := resp.Routes[0]
```
Comment (carlaKC) [inline lightning_client.go:3249]: How does lnd sort routes if there are more than one? Might make sense to always return the cheapest route (perhaps 0 index is already that?)

**Review by carlaKC — COMMENTED** [review]
Looking good! Why is this "incomplete", because we only return one route?

File: lightning_client.go (hunk 2 above)
Comment (bhandras) [reply lightning_client.go:828]: Hop hints are used to simply extend the graph and while the router knows about them, inclusion in the final route is not guaranteed.

File: lightning_client.go (hunk 4 above)
Comment (bhandras) [reply lightning_client.go:3249]: LND always returns one route, the `repeated` property in the proto is just there for historical reasons as far as I can see.

**Comment by bhandras** [comment]
> Looking good! Why is this "incomplete", because we only return one route?

Incomplete because a bunch of things are not returned (TLV, MPP records, etc).

File: lightning_client.go (hunk 1 above)
Comment (bhandras) [reply lightning_client.go]: done

File: lightning_client.go (hunk 3 above)
Comment (bhandras) [reply lightning_client.go]: "Repurposed" an error for this as it seems there were some unused one in the file.

**Review by guggero — APPROVED** [review]
LGTM :tada:

File: lightning_client.go (hunk 4 above)
Comment (carlaKC) [reply lightning_client.go:3243]: Ah I see, TIL.

**Review by carlaKC — APPROVED** [review]
🌟

**Pull request merged**